### PR TITLE
Split CST and derivings

### PR DESCRIPTION
**Pull request by Niols**
In the PR, we split the module containing the CST and the modules containing serialisers for the CST (to and from JSON) or the visitors for the CST.

Ideally, this would be done using something OCamly, like ppx_import. Sadly, it seems that it isn't maintained so much anymore and, anyway, it uses technologies that clash with the Dune way of building PPXes (used by visitors and ppx_deriving in the project). I am not aware of any other good solution for that. I therefore tried an approach based on the copy of the `CST.ml` file (which now only contains types and has been renamed `CST.mli` as well as some text manipulations. I am not necessarily pleased by this solution but it does seem to work. I would add that problems are mitigated by the fact that we compile the result of the text manipulations, which prevents us from breaking everything.

I believe it makes for a cleaner architecture of the Morbig library, because the CST module only contains types (as the name suggests) and the serialisers and visitors are hanging out in appropriately named modules. Additionally, it makes for a cleaner documentation and allows us to not build documentation for the visitors (which is unreadable anyways), therefore fixing #124 (the size of the documentation goes from ~500M to ~5M).

I keep this PR as a draft for now as I am pretty sure that the script would not work on Windows. I will test that shortly, once I fix the CI.

In the meantime, @yurug, do you have anything to say? Does the solution look reasonable to you or is it awful and should be avoided? Do you know of any other way to do the same thing?

### CODE REVIEW

**Comment by Niols** [comment]
I have now introduced #128 which is based on this PR but rewrites the splitter in OCaml instead of Shell. This does not make the compilation faster or clearer. It does allow for a better documentation of the process (in particular, the regexp can be split and documented in tiny bits). More importantly, it fixes the compilation on Windows, which we will see once we integrate the new CI (see #127).

We have to choose between this PR, #128 or none of the two if we don't validate the concept. @yurug, WDYT?

File: src/generate-CST-derivings.sh [hunk 1]
```diff
@@ -0,0 +1,125 @@
+#!/bin/sh
+############################################################################
+##  Copyright (C) 2017-2021 Yann Régis-Gianas, Nicolas Jeannerod,         ##
```
Comment (yurug) [inline src/generate-CST-derivings.sh:3]: Did you `shellcheck` this script shell?

File: src/generate-CST-derivings.sh [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+##     and foo = $1.foo = ...
+##     and foo = $1.foo = ...
+##
+## where $1 is the first argument, except when what follows the equal sign
+## starts with a lowercase, because this would then be a syntax equality and we
+## must not add an other one. This function is useful to copy a whole type while
+## keeping it equal with the original module's type.
+
+add_type_equalities () {
+    sed "s/\(type\|and\)\(.* \)\([a-zA-Z_'][a-zA-Z_']*\) *=\([^=]*=\)\? *\([^ a-z]\|$\)/\1\2\3 = \2CST.\3 = \5/g"
```
Comment (yurug) [inline src/generate-CST-derivings.sh:103]: If I understand correctly, you are specifying the language for type identifiers in the first occurrence (right after the `type`) while the type identifier between the two equals is less constrained. Any reason for that? 

(I am not asking for a change, just curious.)

File: src/generate-CST-derivings.sh (hunk 1 above)
Comment (Niols) [reply src/generate-CST-derivings.sh:3]: I did! (And I have ShellCheck on the fly in my editor.) The only thing it complains about is my use of `cat` line 109; and I do agree with ShellCheck that this `cat` is useless but I believe it improves clarity of the code.

File: src/generate-CST-derivings.sh (hunk 2 above)
Comment (Niols) [reply src/generate-CST-derivings.sh:103]: No really good reason for that, no, it just took me a while to write this regular expression and I might have changed my strategy in the middle. I think I like the OCaml version (of #128, which I wrote afterwards) better.

**Comment by Niols** [comment]
Closed in favour of #128.

**Pull request closed without merging**